src/typeof.cpp

```diff
@@ -14,58 +14,130 @@
-  }
-  inline void destruct(void* ptr) const {
-    reinterpret_cast<null*>(ptr)->~null();
+// TypeInfos owns all the dap::TypeInfo instances.
+struct TypeInfos {
+  // get() returns the TypeInfos singleton pointer.
+  // TypeInfos is constructed with an internal reference count of 1.
+  static TypeInfos* get();
+
+  // reference() increments the TypeInfos reference count.
+  inline void reference() {
+    assert(refcount.load() > 0);
+    refcount++;
   }
-  inline bool deserialize(const dap::Deserializer*, void*) const {
-    return true;
+
+  // release() decrements the TypeInfos reference count.
+  // If the reference count becomes 0, then the TypeInfos is destructed.
+  inline void release() {
+    if (--refcount == 0) {
+      this->~TypeInfos();
+    }
   }
-  inline bool serialize(dap::Serializer*, const void*) const { return true; }
+
+  struct NullTI : public dap::TypeInfo {
+    using null = dap::null;
+    inline std::string name() const { return "null"; }
+    inline size_t size() const { return sizeof(null); }
+    inline size_t alignment() const { return alignof(null); }
+    inline void construct(void* ptr) const { new (ptr) null(); }
+    inline void copyConstruct(void* dst, const void* src) const {
+      new (dst) null(*reinterpret_cast<const null*>(src));
+    }
+    inline void destruct(void* ptr) const {
+      reinterpret_cast<null*>(ptr)->~null();
+    }
+    inline bool deserialize(const dap::Deserializer*, void*) const {
+      return true;
+    }
+    inline bool serialize(dap::Serializer*, const void*) const { return true; }
+  };
+
+  dap::BasicTypeInfo<dap::boolean> boolean = {"boolean"};
+  dap::BasicTypeInfo<dap::string> string = {"string"};
+  dap::BasicTypeInfo<dap::integer> integer = {"integer"};
+  dap::BasicTypeInfo<dap::number> number = {"number"};
+  dap::BasicTypeInfo<dap::object> object = {"object"};
+  dap::BasicTypeInfo<dap::any> any = {"any"};
+  NullTI null;
+  std::vector<std::unique_ptr<dap::TypeInfo>> types;
+
+ private:
+  TypeInfos() = default;
+  ~TypeInfos() = default;
+  std::atomic<uint64_t> refcount = {1};
+};
+
+// aligned_storage() is a replacement for std::aligned_storage that isn't busted
+// on older versions of MSVC.
+template <size_t SIZE, size_t ALIGNMENT>
+struct aligned_storage {
+  struct alignas(ALIGNMENT) type {
+    unsigned char data[SIZE];
+  };
 };
 
-static dap::BasicTypeInfo<dap::boolean> booleanTI("boolean");
-static dap::BasicTypeInfo<dap::string> stringTI("string");
-static dap::BasicTypeInfo<dap::integer> integerTI("integer");
-static dap::BasicTypeInfo<dap::number> numberTI("number");
-static dap::BasicTypeInfo<dap::object> objectTI("object");
-static dap::BasicTypeInfo<dap::any> anyTI("any");
-static NullTI nullTI;
+TypeInfos* TypeInfos::get() {
+  static aligned_storage<sizeof(TypeInfos), alignof(TypeInfos)>::type memory;
+
+  struct Instance {
+    aligned_storage<sizeof(TypeInfos), alignof(TypeInfos)>::type ti;
```

Comment:
aligned_storage<sizeof(TypeInfos), alignof(TypeInfos)>::type ti;
Usage was not found. Why do we need it?

Reply:
We don't. Good spot. Removed.